ultralytics/engine/trainer.py

```diff
@@ -475,6 +477,16 @@ def _do_train(self, world_size=1):
         torch.cuda.empty_cache()
         self.run_callbacks("teardown")
 
+    def _model_train(self):
+        """Set model in training mode."""
+        self.model.train()
+
+        # Freeze BN
+        if self.args.freeze_bn:
+            for n, m in self.model.named_modules():
+                if any(x in n for x in self.freeze_layer_names) and isinstance(m, nn.BatchNorm2d):
```

Comment:
@caopulan since I was looking into something that ended up being related to this PR, I was messing around with the `modules` and found that using this was a bit faster when I was testing in isolation

```py
for n, m in self.model.named_modules():
        if any(filter(lambda f: f in n, self.freeze_layer_names)) and '.bn' in n:
            m.eval()
```

I was doing other things while the tests ran, so there was fluctuations each time. The average of each version were `3.03 (s)` when using `filter` and `3.23 (s)` when using the line from the PR.

Nothing critical, just happened to be here again so figured I'd share in case you wanted to test it out. 

<details><summary>Speed test isolated code</summary>
<p>

```py

import timeit
from ultralytics import YOLO

def m_init() -> tuple[YOLO, list[int], list[str]]:
   """initialize speed-test components"""
    model = YOLO("yolov8n.pt")
    _list = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    names = [f"model.{x}." for x in freeze_list]
    return model, _list, names

def bn_eval() -> None:
    """used to check speed"""
    detect, freeze_layer_names, freeze_layer_names = m_init()
    
    for n, m in detect.named_modules():
        #if any(x in n for x in self.freeze_layer_names) and isinstance(m, nn.BatchNorm2d):
        if any(filter(lambda f: f in n, freeze_layer_names)) and '.bn' in n:
            m.eval()

timeit.repeat(bn_eval, number=100, repeat=5)

```

</p>
</details>

Reply:
I think using `filter` is reasonable but checking modules by their names may cause unexpected problems. For example, if a BN layer named by `batch_norm' or `conv_bn` in the future would make it ineffective. Hence I suggest to maintain `isinstance(m, nn.BatchNorm2d)` and change `any(x in n for x in self.freeze_layer_names)` to `any(filter(lambda f: f in n, freeze_layer_names))`. Less than 1 second delay per epoch is acceptable, I think.
What do you think?

Reply:
I think that's totally reasonable 🎉

Reply:
OK, i'll update it.